shared/src/main/scala/io/lemonlabs/uri/Uri.scala

```diff
@@ -495,6 +496,45 @@ sealed trait Url extends Uri {
     case _ =>
       false
   }
+
+  /** @return the reference URL resolved with this URL as the base according to section 5.2.2 Transform references of
+    * <a href="http://www.ietf.org/rfc/rfc3986.txt">RFC 3986</a>.
+    */
+  def resolve(ref: Url, strict: Boolean = false): Url = {
+    val baseScheme = schemeOption match {
+      case Some(scheme) => scheme
+      case None         => throw new UriException("Base URI for resolution must have a scheme")
+    }
+    val refScheme = ref.schemeOption match {
+      case Some(scheme) if !strict && scheme == baseScheme => None
+      case mScheme                                         => mScheme
+    }
+
+    if (refScheme.isDefined) {
+      ref.removeDotSegments
+    } else if (ref.isInstanceOf[UrlWithAuthority]) {
```

Comment:
Does `authorityOption.nonEmpty` work here instead of the `isInstanceOf` check? I think that would be nicer